truffle/src/com.oracle.truffle.polyglot/src/com/oracle/truffle/polyglot/PolyglotLanguageContext.java

```diff
@@ -619,69 +619,81 @@ void ensureCreated(PolyglotLanguage accessingLanguage) {
-                        }
-                        created = true;
-                    } finally {
-                        if (!wasCreated && eventsEnabled) {
-                            EngineAccessor.INSTRUMENT.notifyLanguageContextCreateFailed(context.engine, context.creatorTruffleContext, language.info);
+
+                            // no more errors after this line
+                            creatingThread = Thread.currentThread();
+                            env = localEnv;
+                            lazy = localLazy;
+                            assert EngineAccessor.LANGUAGE.getLanguage(env) != null;
+
+                            try {
+                                List<Object> languageServicesCollector = new ArrayList<>();
+                                Object contextImpl = LANGUAGE.createEnvContext(localEnv, languageServicesCollector);
+                                language.initializeContextClass(contextImpl);
+                                String errorMessage = verifyServices(language.info, languageServicesCollector, language.cache.getServices());
+                                if (errorMessage != null) {
+                                    throw PolyglotEngineException.illegalState(errorMessage);
+                                }
+                                PolyglotFastThreadLocals.notifyLanguageCreated(this);
+                                this.languageServices = languageServicesCollector;
+                                if (language.isHost()) {
+                                    context.initializeHostContext(this, context.config);
+                                }
+                                wasCreated = true;
+                                if (eventsEnabled) {
+                                    EngineAccessor.INSTRUMENT.notifyLanguageContextCreated(context.engine, context.creatorTruffleContext, language.info);
+                                }
+                                context.invokeContextLocalsFactory(context.contextLocals, languageInstance.contextLocalLocations);
+                                context.invokeContextThreadLocalFactory(languageInstance.contextThreadLocalLocations);
+
+                                languageInstance = null; // commit language use
+                            } catch (Throwable e) {
+                                env = null;
+                                lazy = null;
+                                throw e;
+                            } finally {
+                                creatingThread = null;
+                            }
+                            created = true;
                         }
                     }
+                    break LOOP;
+                }
+            } finally {
+                if (!wasCreated && eventsEnabled) {
+                    EngineAccessor.INSTRUMENT.notifyLanguageContextCreateFailed(context.engine, context.creatorTruffleContext, language.info);
```

Comment:
Please address my above comment. The behavior should be exactly the same if the threadDeniedAccessHandler is not set. Don't move the finally block outside of the synchronized block. You can put `threadAccessException == null` condition into the finally block or something, so that the LanguageContextCreateFailed event is not executed multiple times in case of retries. Moreover, your current solution is firing the LanguageContextCreateFailed event even if the language context is already created, which could be easily fixed by setting the `wasCreated` to true in that case, but instead please do the larger suggested change.

Reply:
Great catch! Addressed in b8e8590caa5.

The goal is to cause _no changes_ to the current behavior of the code except looping. According to diff in my IDE that hides the whitespace changes:

![diff](https://github.com/user-attachments/assets/965a48bc-dd77-44f6-b687-97d7d97a56a2)

We only wrap the existing code with the `loop: for (;;) {` and don't change any other behavior since b8e8590caa5.

Reply:
Thanks, I am also fine with triggering the notifyLanguageContextCreate and notifyLanguageContextCreateFailed multiple times in case of retries.